ngramoffset.go

```diff
@@ -112,3 +122,211 @@ func (a *arrayNgramOffset) DumpMap() map[ngram]simpleSection {
 func (a *arrayNgramOffset) SizeBytes() int {
 	return 8*len(a.tops) + 4*len(a.bots) + 4*len(a.offsets)
 }
+
+// combinedNgramOffset combines an ascii ngram mapping with a unicode ngram mapping,
+// falling back on unicode for unicode ngrams or ascii ngrams with excessive lengths.
+type combinedNgramOffset struct {
+	asc *asciiNgramOffset
+	uni *arrayNgramOffset
+}
+
+func makeCombinedNgramOffset(ngrams []ngram, offsets []uint32) combinedNgramOffset {
+	// split ngrams & offsets into ascii ngrams and unicode ngrams,
+	// since ascii ngrams can be represented much more compactly (21b instead of 63b)
+	ngramsAscii := make([]uint32, 0, len(ngrams)*11/10)
+	offsetsAscii := make([]uint32, 0, len(ngrams)*11/10)
+
+	ngramsUnicode := make([]ngram, 0, len(ngrams)*2/10)
+	offsetsUnicode := make([]uint32, 0, len(ngrams)*2/10)
+
+	for i, ng := range ngrams {
+		if ng&ngramAsciiMask == ng {
+			ngp := ngramAsciiToPacked(ng)
+			if i == len(ngrams)-1 || ngrams[i+1]&ngramAsciiMask != ngrams[i+1] {
+				// at the end of a section we insert an extra offset with the same ngram,
+				// so the size of the segment can be calculated properly
+				ngramsAscii = append(ngramsAscii, ngp, ngp)
+				offsetsAscii = append(offsetsAscii, offsets[i], offsets[i+1])
+			} else {
+				ngramsAscii = append(ngramsAscii, ngp)
+				offsetsAscii = append(offsetsAscii, offsets[i])
+			}
+			if offsets[i+1]-offsets[i] >= ngramAsciiMaxSectionLength {
+				// max-length ascii sections can't be represented properly in the ascii mapping,
+				// and are duplicated in the normal unicode entries.
+				ngramsUnicode = append(ngramsUnicode, ng, ng)
+				offsetsUnicode = append(offsetsUnicode, offsets[i], offsets[i+1])
+			}
+		} else {
+			if i == len(ngrams)-1 || ngrams[i+1]&ngramAsciiMask == ngrams[i+1] {
+				ngramsUnicode = append(ngramsUnicode, ng, ng)
+				offsetsUnicode = append(offsetsUnicode, offsets[i], offsets[i+1])
+			} else {
+				ngramsUnicode = append(ngramsUnicode, ng)
+				offsetsUnicode = append(offsetsUnicode, offsets[i])
+			}
+		}
+	}
+
+	if len(ngramsAscii) > 0 {
+		ngramsAscii = ngramsAscii[:len(ngramsAscii)-1]
+	}
+	if len(ngramsUnicode) > 0 {
+		ngramsUnicode = ngramsUnicode[:len(ngramsUnicode)-1]
+	}
+
+	asc := makeAsciiNgramOffset(ngramsAscii, offsetsAscii)
+	uni := makeArrayNgramOffset(ngramsUnicode, offsetsUnicode)
+
+	return combinedNgramOffset{asc, &uni}
+}
+
+// Get returns a simpleSection with sz=0 if no entry, or sz=ngramAsciiMaxSectionLength
+// if the length of the ngram is too large for this type and it should cascade to the next entry.
```

Comment:
I think this comment is incorrect. You don't return `sz=ngramAsciiMaxSectionLength`. If I am not mistaken this should have the same return values as `arrayNgramOffset` or `map[ngram]simpleSection` when constructed correctly.

Just to make sure I understand, you use `sz=0` or
`sz=ngramAsciiMaxSectionLength` from ascii as a sign to look into ngram map.